lib/matplotlib/_constrained_layout.py

```diff
@@ -83,181 +74,360 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
     decorations on the axis.
+
+    Then make room for colorbars.
+    """
+    for panel in fig.panels:  # recursively make child panel margins
+        ss = panel._subplotspec
+        _make_layout_margins(panel, renderer, w_pad=w_pad, h_pad=h_pad,
+                             hspace=hspace, wspace=wspace)
+
+        margins = _get_margin_from_padding(panel, w_pad=0, h_pad=0,
+                                           hspace=hspace, wspace=wspace)
+        panel._layoutgrid.parent.edit_outer_margin_mins(margins, ss)
+
+    # for ax in [a for a in fig._localaxes if hasattr(a, 'get_subplotspec')]:
+
+    for ax in [a for a in fig.get_axes() if hasattr(a, 'get_subplotspec')]:
+        ss = ax.get_subplotspec()
+        gs = ss.get_gridspec()
+        nrows, ncols = gs.get_geometry()
+
+        if gs._layoutgrid is None:
+            return
+
+        margin = _get_margin_from_padding(ax, w_pad=w_pad, h_pad=h_pad,
+                                           hspace=hspace, wspace=wspace)
+        margin0 = margin.copy()
+        pos, bbox = _get_pos_and_bbox(ax, renderer)
+
+        # the margin is the distance between the bounding box of the axes
+        # and its position (plus the padding from above)
+        margin['left'] += -bbox.x0 + pos.x0
+        margin['right'] += bbox.x1 - pos.x1
+        # remember that rows are ordered from top:
+        margin['bottom'] += pos.y0 - bbox.y0
+        margin['top'] += -pos.y1 + bbox.y1
+
+        # make margin for colorbars.  These margins go in the
+        # padding margin, versus the margin for axes decorators.
+        for cbax in ax._colorbars:
+            # note pad is a fraction of the parent width...
+            pad = _colorbar_get_pad(cbax)
+            # colorbars can be child of more than one subplot spec:
+            cbp_rspan, cbp_cspan = _get_cb_parent_spans(cbax)
+            loc = cbax._colorbar_info['location']
+            if loc == 'right':
+                if cbp_cspan.stop == ss.colspan.stop:
+                    # only increase if the colorbar is on the right edge
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['rightcb'] += cbbbox.width + pad
+            elif loc == 'left':
+                if cbp_cspan.start == ss.colspan.start:
+                    # only increase if the colorbar is on the left edge
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['leftcb'] += cbbbox.width + pad
+            elif loc == 'top':
+                if cbp_rspan.start == ss.rowspan.start:
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['topcb'] += cbbbox.height + pad
+            else:
+                if cbp_rspan.stop == ss.rowspan.stop:
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['bottomcb'] += cbbbox.height + pad
+            # If the colorbars are wider than the parent box in the
```

Comment:
@anntzer: new work to fix overlapping colorbars here.  I wasn't taking into account the cross-direction...

Reply:
That's much better now, but I'd argue the space is too big between the axes and the colorbar?  Going back to a figsize of (8, 5):

old:
![old](https://user-images.githubusercontent.com/1322974/85533951-3ea63280-b611-11ea-884a-e27fc1aa61ce.png)

new:
![new](https://user-images.githubusercontent.com/1322974/85533940-3cdc6f00-b611-11ea-9836-be566133d82c.png)

I guess one should try (if possible) to keep the vspace between the suptitle and the first row of axes at least not smaller than the vspace between axes themselves.  "Sticking" the colorbars as much as possible to the axes would be even nicer, I guess.

Reply:
> That's much better now, but I'd argue the space is too big between the axes and the colorbar? Going back to a figsize of (8, 5):

Right - the new version respects the `pad` default values, which for location = "bottom" which is 0.15 (to make way for the xlabels).  We could change this default if constrained_layout is being used - its just a bit of extra documentation.  

You can of course control the pad with the `pad` kwarg to colorbar:

` fig.colorbar(im, ax=ax, orientation="horizontal", pad=0.05)`

![colorbarissues](https://user-images.githubusercontent.com/1562854/85582984-ba44b700-b5f2-11ea-9206-df916a2c546a.png)

Reply:
OK< changed the default to 0.05 for all constrained_layout axes.  See new line in `colorbar.py` and new documentation of *pad* kwarg.  I agree that's better and should be the default.